libraries/write_chef_error_handler.rb

```diff
@@ -0,0 +1,76 @@
+      # if the error file already exists we don't take any additional action here
+      unless File.exist?(error_file)
+        message_error = 'Failed to run chef recipe.'
+        node_type_to_log = {
+          'HeadNode' => 'Please check /var/log/chef-client.log in the head node, or check the chef-client.log in CloudWatch logs.',
+          'ComputeFleet' => 'Please check the cloud-init-output.log in CloudWatch logs.',
+        }
+        message_logs_to_check = node_type_to_log[node['cluster']['node_type']]
+        message_troubleshooting_link = 'Please refer to'\
+          ' https://docs.aws.amazon.com/parallelcluster/latest/ug/troubleshooting-v3.html#troubleshooting-v3-get-logs'\
+          ' for more details on ParallelCluster logs.'
+
+        # get the failed action records using the chef function filtered_collection
+        # reference: https://github.com/cinc-project/chef/blob/stable/cinc/lib/chef/action_collection.rb#L107
+        failed_action_collection = action_collection.filtered_collection(
+          up_to_date: false, skipped: false, updated: false, failed: true, unprocessed: false
+        )
+
+        # define a mapping from the mount-related resource name to the error message we would like to display
+        mount_message_mapping = {
+          "add ebs" => "Failed to mount EBS volume.",
+          "add raid" => "Failed to mount RAID array.",
+          "mount efs" => "Failed to mount EFS.",
+          "mount fsx" => "Failed to mount FSX.",
+        }
+
+        failed_action_collection.each do |action_record|
+          # there might be multiple failed action records
+          # here we only look at the outermost layer resource by setting nesting_level = 0
+          # with the assumption that there is only one failed action record with nesting_level = 0
+          next unless action_record.nesting_level == 0
+          # we first check if it is a storage mounting error for EBS, RAID, EFS, or FSX,
+          # otherwise we will get the recipe information
+          message_error = mount_message_mapping[action_record.new_resource.name] || "Failed to run chef recipe#{get_recipe_info(action_record)}."
+          break
+        end
+
+        # at the end, put together and store the full error message into the dedicated file
+        shell_out("echo '#{message_error} #{message_logs_to_check} #{message_troubleshooting_link}'> '#{error_file}'")
+
+      end
+    end
+
+    def get_recipe_info(action_record)
+      # use the built-in function "defined_at" of Chef::Resource to get the recipe information
+      # when source_line is not available it will return "dynamically defined" and we replace it with empty string
+      # reference: https://github.com/cinc-project/chef/blob/stable/cinc/lib/chef/resource.rb#L1436
+      recipe_info = action_record.new_resource.defined_at
+      recipe_info == "dynamically defined" ? "" : recipe_info
```

Comment:
Isn't it possible to at least return the recipe name even if the precise failing source line is not available?

Reply:
According to Chef codebase: 

- The `source_line` is the caller of the resource: https://github.com/cinc-project/chef/blob/stable/cinc/lib/chef/dsl/declare_resource.rb#L263 (It's usually available)

- A resource is considered as "dynamic resource" when `source_line` is null:
https://github.com/cinc-project/chef/blob/stable/cinc/lib/chef/formatters/error_inspectors/resource_failure_inspector.rb#L98 

- A resource usually falls into the definition of "Dynamically defined resource" when the cookbook and recipe name are not available. Reference: 
https://github.com/cinc-project/chef/blob/stable/cinc/lib/chef/formatters/doc.rb#L227

- All of the above information and the logic of the `defined_at` built-in method seems to assume that when `source_line` is not available, recipe information will also not be available anyway.

This is consistent with our observation that we've seen situations when we have `source_line` but not recipe name, but we never see any situation when the `source_line` is missing but the recipe name is available. Based on this I think we can keep use the `defined_at` method to avoid the code to be too complicated. Worst case scenario (in some very rare corner case and we can dig into it later), the returned error message is just the general catch-all message. Does that sound good to you?